### ISSUE TITLE
Exception after upgrading the EF Core package to 2.0.1

### ISSUE THREAD

**Issue by andreluizsecco** [dotnet/efcore#10360]
My project was working perfectly. I updated the EF Core packages to 2.0.1 and now I get the following exception:

> fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HL9GGHE4EU10", Request id "0HL9GGHE4EU10:00000001": An unhandled exception was thrown by the application.
System.InvalidOperationException: The entity type 'ProductPlanCustomer' requires a primary key to be defined.
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelValidator.ValidateNonNullPrimaryKeys(IModel model)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelValidator.Validate(IModel model)
   at Microsoft.EntityFrameworkCore.Infrastructure.RelationalModelValidator.Validate(IModel model)
   at Microsoft.EntityFrameworkCore.Internal.SqlServerModelValidator.Validate(IModel model)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.CreateModel(DbContext context, IConventionSetBuilder conventionSetBuilder, IModelValidator validator)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.<>c__DisplayClass5_0.<GetModel>b__0(Object k)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.GetModel(DbContext context, IConventionSetBuilder conventionSetBuilder, IModelValidator validator)
   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.CreateModel()
   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.get_Model()
   at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkServicesBuilder.<>c.<TryAddCoreServices>b__7_1(IServiceProvider p)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.<>c__DisplayClass22_0.<RealizeService>b__0(ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.EntityFrameworkCore.DbContext.get_DbContextDependencies()
   at Microsoft.EntityFrameworkCore.DbContext.get_InternalServiceProvider()
   at Microsoft.EntityFrameworkCore.DbContext.get_DbContextDependencies()
   at Microsoft.EntityFrameworkCore.DbContext.Set[TEntity]()

> 

Class ProductPlanCustomer:

```c#
public class ProductPlanCustomer
    {
        public int ProductPlanID { get; set; }
        public int CustomerID { get; set; }
        public DateTime HireDate { get; set; }
        public DateTime? CancellationDate { get; set; }
        public byte DueDay { get; set; }
        public decimal? Price { get; set; }

        public virtual ProductPlan ProductPlan { get; set; }
        public virtual Customer Customer { get; set; }
    }
```

Context:

```c#
protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
 {
            var config =  new ConfigurationBuilder()
                .SetBasePath(Directory.GetCurrentDirectory())
                .AddJsonFile("appsettings.json")
                .Build();
            
            optionsBuilder.UseSqlServer(config.GetConnectionString("DefaultConnection"));
}

protected override void OnModelCreating(ModelBuilder modelBuilder)
{
            modelBuilder.ApplyConfiguration(new ProductPlanCustomerMap());

            base.OnModelCreating(modelBuilder);
}
```
Entity Mapping:

```c#
public class ProductPlanCustomerMap : IEntityTypeConfiguration<ProductPlanCustomer>
    {
        public void Configure(EntityTypeBuilder<ProductPlanCustomer> builder)
        {
            builder.ToTable("ProductPlanCustomer");

            builder.HasKey(p => new { p.ProductPlanID, p.CustomerID });

            builder.Property(p => p.HireDate)
                .HasColumnType("date");

             builder.Property(p => p.CancellationDate)
                .HasColumnType("date");

            builder.Property(p => p.Price)
                .HasColumnType("money");

            builder.Property<DateTime>("LastUpdated");

            builder.HasOne(p => p.ProductPlan)
                .WithMany()
                .HasForeignKey(p => p.ProductPlanID);

            builder.HasOne(p => p.Customer)
                .WithMany(p => p.ProductPlans)
                .HasForeignKey(p => p.CustomerID);
        }
    }
```

Any idea of the cause of the exception?

**Comment by ralmsdeveloper** [comment]
Hello André,
You could post your ProductPlan and Customer to try repro.

More for now, make sure that Microsoft.AspNetCore.All is in version 2.0.3

**Comment by andreluizsecco** [comment]
Hello, I'm using the Microsoft.AspNetCore.All package in version 2.0.3

ProductPlan:

```c#
public class ProductPlan
    {
        public int ID { get; set; }
        public string Name { get; set; }
        public string Description { get; set; }
        public decimal Price { get; set; }
        public bool Active { get; set; }
    }
```

Customer:

```c#
public class Customer
    {
        public int ID { get; set; }
        public string Name { get; set; }
        public string Sex { get; set; }
        public long? Phone { get; set; }
        public string Email { get; set; }

        public virtual ICollection<ProductPlanCustomer> ProductPlans { get; set; }
    }
```

**Comment by ralmsdeveloper** [comment]
Hello André,

I performed an insert with the instructions below and I was successful!

```csharp
using (var db = new Issue10360Context())
{
    db.Database.EnsureDeleted();
    db.Database.EnsureCreated(); 

    db.ProductPlanCustomer.Add(
        new ProductPlanCustomer
    {
        Price = 2,
        CancellationDate = DateTime.Now,
        HireDate = DateTime.Now,
        Customer = new Customer
        {
            Email = "andre@andre.xxx",
            Name = "Andre"
        },
        ProductPlan = new ProductPlan
        {
            Active = true,
            Description = "Test"
        }
    });
    db.SaveChanges(); 

}
```

From what I realized you have keys that values need to be propagated, here we had something similar.
https://github.com/aspnet/EntityFrameworkCore/issues/9552#issuecomment-324794136

I believe @ajcvickers or @smitpatel could complement this.

**Comment by ajcvickers** [comment]
@andreluizsecco I put your code in a console app and it is working fine for me--see below. Can you either update the code below or post a complete, runnable project that reproduces what you are seeing?

```C#
public class ProductPlanCustomer
{
    public int ProductPlanID { get; set; }
    public int CustomerID { get; set; }
    public DateTime HireDate { get; set; }
    public DateTime? CancellationDate { get; set; }
    public byte DueDay { get; set; }
    public decimal? Price { get; set; }

    public virtual ProductPlan ProductPlan { get; set; }
    public virtual Customer Customer { get; set; }
}

public class ProductPlan
{
    public int ID { get; set; }
    public string Name { get; set; }
    public string Description { get; set; }
    public decimal Price { get; set; }
    public bool Active { get; set; }
}

public class Customer
{
    public int ID { get; set; }
    public string Name { get; set; }
    public string Sex { get; set; }
    public long? Phone { get; set; }
    public string Email { get; set; }

    public virtual ICollection<ProductPlanCustomer> ProductPlans { get; set; }
}

public class ProductPlanCustomerMap : IEntityTypeConfiguration<ProductPlanCustomer>
{
    public void Configure(EntityTypeBuilder<ProductPlanCustomer> builder)
    {
        builder.ToTable("ProductPlanCustomer");

        builder.HasKey(p => new { p.ProductPlanID, p.CustomerID });

        builder.Property(p => p.HireDate)
            .HasColumnType("date");

        builder.Property(p => p.CancellationDate)
            .HasColumnType("date");

        builder.Property(p => p.Price)
            .HasColumnType("money");

        builder.Property<DateTime>("LastUpdated");

        builder.HasOne(p => p.ProductPlan)
            .WithMany()
            .HasForeignKey(p => p.ProductPlanID);

        builder.HasOne(p => p.Customer)
            .WithMany(p => p.ProductPlans)
            .HasForeignKey(p => p.CustomerID);
    }
}

public class TestDbContext : DbContext
{
    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.ApplyConfiguration(new ProductPlanCustomerMap());

        base.OnModelCreating(modelBuilder);
    }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder.UseSqlServer(
            @"Server=(localdb)\mssqllocaldb;Database=Test;ConnectRetryCount=0");
}

public class Program
{
    public static void Main()
    {
        using (var context = new TestDbContext())
        {
            context.Database.EnsureDeleted();
            context.Database.EnsureCreated();

            context.Add(
                new ProductPlanCustomer
                {
                    ProductPlan = new ProductPlan(),
                    Customer = new Customer()
                });

            context.SaveChanges();
        }
    }
}
```

**Comment by andreluizsecco** [comment]
Checking out here, I noticed that some parts are working, like insert, update and delete directly on this table.
The exception happens in this method:

```c#
public virtual bool IsValidTenant(int? customerID, string userID)
{
     return db.Set<CustomerUser>()
         .Any(x => x.CustomerID == customerID && x.UserID.Equals(userID));
}
```

Structure and mapping of 'CustomerUser':

```c#
public class CustomerUser
    {
        public int ID { get; set; }
        public int CustomerID { get; set; }
        public string UserID { get; set; }

        public virtual Customer Customer { get; set; }
        public virtual User User { get; set; }
    }


public class CustomerUserMap : IEntityTypeConfiguration<CustomerUser>
    {
        public void Configure(EntityTypeBuilder<CustomerUser> builder)
        {
            builder.ToTable("CustomerUser");

            builder.Property(p => p.ID)
                .ValueGeneratedOnAdd();

            builder.HasOne(p => p.Customer)
                .WithMany(p => p.Users)
                .HasForeignKey(p => p.CustomerID);

            builder.HasOne(p => p.User)
                .WithMany()
                .HasForeignKey(p => p.UserID);
        }
    }
```
Obs: The 'User' class is created by my authentication mechanism, I am using ASP.NET Core Identity 2.0.1

**Comment by andreluizsecco** [comment]
I discovered the problem. It's an important detail that I forgot to mention. The context I am using to running the query in the CustomerUser class is not the same as where I am applying the Mapping of the ProductPlanCustomer class, so it is not achieved to determine the primary key. To perform a test, I have added the following line in the OnConfiguring method of the context that I am using and it worked:

```c#
modelBuilder.Entity<ProductPlanCustomer>().HasKey(p => new { p.ProductPlanID, p.CustomerID });
```

Now the mistake seems to make a lot of sense to me. I'm just wondering why with the packages for EF Core in version 2.0.0 it's working (running in production)

Any thoughts on this?
